### ISSUE TITLE
vcs export --exact reports tag names instead of commit hash

### ISSUE THREAD

**Issue by crdelsey** [dirk-thomas/vcstool#109]
**Version:** vcs 0.2.4

**How to Reproduce:**
Run `vcs export --exact` on a set of repos.

**Expected behavior:**
The version field of each repo will report the exact commit hash like this:
```
  src/ros2/rclpy:
    type: git
    url: https://github.com/ros2/rclpy.git
    version: 76d26a125da7f9df19ba35beb60cb16838019a10
```

**Actual Behavior:**
Frequently the hash is replaced by a tag name.
```
  src/ros2/rcl_interfaces:
    type: git
    url: https://github.com/ros2/rcl_interfaces.git
    version: 0.8.0
```

---

I think when the --exact parameter is specified, it is preferable to provide the commit hash.

**Comment by dirk-thomas** [comment]
> I think when the --exact parameter is specified, it is preferable to provide the commit hash.

The behavior to return tags when available is intentional (see #98). Since tags are intended to be immutable and significantly more readable they are being used by `--exact` if the current hash has exactly one tag pointing to it.

**Comment by crdelsey** [comment]
I don't know about all vcs tools, but tags frequently aren't all that immutable. For instance, they are frequently used to mark the `latest_stable` version or some such thing, with the tag being moved as someone or CI decides that a new version is now passing the tests.

I'm pretty sure they can be moved in Git, so, for right or wrong, if somebody moves the tag, the --exact option is not so exact anymore.

Having the tag is a nice feature most of the time, but, it disagrees with the docs for the --exact parameter.
```
"export" command parameters:
  --exact            Export exact commit hash instead of branch names
```

**Comment by dirk-thomas** [comment]
> I'm pretty sure they can be moved in Git, so, for right or wrong, if somebody moves the tag, the --exact option is not so exact anymore.

Sure, you can also rewrite history in git - still something to consider not doing ;-)

> Having the tag is a nice feature most of the time, but, it disagrees with the docs for the --exact parameter.

Agreed, clarifying the help text would certainly be good :+1:

> If you want to keep this behavior, maybe another option, something like --hash-only could be used to disable the tag lookup feature?

Adding another option to toggle the behavior to always return hashes vs. potentially tags if they are unique would be good. Since the change has only landed recently I am not too concerned about changing the behavior (if we think that makes more sense). I am just not sure what the right names for the two options would be `--exact` and ? `--hash-only` seems very git specific...

**Comment by dirk-thomas** [comment]
After thinking about it I think it is more natural if `--exact` always outputs hashes and a new option `--exact-with-tags` (naming up for discussion) for the current behavior (unique tags when available, fallback to hashes).

The new option could either be standalone or require also to pass `--exact`. The former seems more user friendly but will require all existing backends to check for the new option and at least do the same as for the existing option.

@crdelsey Would you be willing to contribute a pull request for this new option?

**Comment by crdelsey** [comment]
Yeah. I can take a stab at it. Probably next week.

I haven't looked at the codebase much yet, so I'm not sure if this makes sense. 

But rather than changing all the backends, I could have the command line parser split the `--exact-with-tags` convenience option into two separate options on the command object: the existing `.exact` and a `.with_tags`. All the other backends would ignore the `with_tags` flag and maintain the same behavior.

**Comment by dirk-thomas** [comment]
@crdelsey Friendly ping.

**Comment by dirk-thomas** [comment]
Resolved by #115.
